### PR TITLE
Op stack hook deployment

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

 Hook deployer includes deployment of `OpStackHook` and `OpStackIsm`

### Drive-by changes

- fixes https://github.com/hyperlane-xyz/issues/issues/622

### Backward compatibility

Yes

### Testing

None

### CODE REVIEW

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts [hunk 1]
```diff
@@ -124,4 +131,38 @@ export class HyperlaneHookDeployer extends HyperlaneDeployer<
     );
     return hooks;
   }
+
+  async deployOpStack(
+    chain: ChainName,
+    config: OpStackHookConfig,
+    coreAddresses = this.core[chain],
+  ): Promise<HyperlaneContracts<HookFactories>[HookType.OP_STACK]> {
```
Comment (yorhodes) [inline typescript/sdk/src/hook/HyperlaneHookDeployer.ts:139]:
```suggestion
  ): Promise<OpStackHook> {
```
probably also works

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts [hunk 2]
```diff
@@ -124,4 +131,38 @@ export class HyperlaneHookDeployer extends HyperlaneDeployer<
...
+    const mailbox = coreAddresses.mailbox;
+    if (!mailbox) {
+      throw new Error(`Mailbox address is required for ${config.type}`);
+    }
```
Comment (yorhodes) [inline typescript/sdk/src/hook/HyperlaneHookDeployer.ts:143]: is this just a sanity check? iiuc this isn't optional

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts [hunk 3]
```diff
@@ -124,4 +131,38 @@ export class HyperlaneHookDeployer extends HyperlaneDeployer<
...
+    const l2Messenger: Address = await l1Messenger.OTHER_MESSENGER();
+    // deploy opstack ism
+    const ismConfig: OpStackIsmConfig = {
+      type: ModuleType.OP_STACK,
+      nativeBridge: l2Messenger,
+    };
+    const opstackIsm = await this.ismFactory.deploy(
+      config.destinationChain,
+      ismConfig,
+    );
```
Comment (yorhodes) [inline typescript/sdk/src/hook/HyperlaneHookDeployer.ts:158]: let's persist the artifacts here?

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts [hunk 4]
```diff
@@ -124,4 +131,38 @@ export class HyperlaneHookDeployer extends HyperlaneDeployer<
...
+    const l1Messenger = IL1CrossDomainMessenger__factory.connect(
+      config.nativeBridge,
+      this.multiProvider.getSignerOrProvider(chain),
+    );
+    const l2Messenger: Address = await l1Messenger.OTHER_MESSENGER();
+    // deploy opstack ism
+    const ismConfig: OpStackIsmConfig = {
+      type: ModuleType.OP_STACK,
+      nativeBridge: l2Messenger,
+    };
```
Comment (yorhodes) [inline typescript/sdk/src/hook/HyperlaneHookDeployer.ts:154]: doesn't the ISM need the hook address?

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts [hunk 5]
```diff
@@ -124,4 +131,38 @@ export class HyperlaneHookDeployer extends HyperlaneDeployer<
...
+      ismConfig,
+    );
+    // deploy opstack hook
+    const hooks = await this.deployContract(chain, HookType.OP_STACK, [
+      mailbox,
+      config.destinationDomain,
+      opstackIsm.address,
+      config.nativeBridge,
+    ]);
+    return hooks;
```
Comment (yorhodes) [inline typescript/sdk/src/hook/HyperlaneHookDeployer.ts:166]: nit
```suggestion
    const hook = await this.deployContract(chain, HookType.OP_STACK, [
      mailbox,
      config.destinationDomain,
      opstackIsm.address,
      config.nativeBridge,
    ]);
    return hook;
```

File: typescript/sdk/src/hook/types.ts [hunk 6]
```diff
@@ -3,12 +3,14 @@ import { BigNumber } from 'ethers';
...
 
 import { IgpConfig } from '../gas/types';
+import { ChainName } from '../types';
 
 export enum HookType {
   MERKLE_TREE = 'merkleTreeHook',
   INTERCHAIN_GAS_PAYMASTER = 'interchainGasPaymaster',
   AGGREGATION = 'aggregationHook',
   PROTOCOL_FEE = 'protocolFee',
+  OP_STACK = 'opStack',
```
Comment (yorhodes) [inline typescript/sdk/src/hook/types.ts:13]: we use this for the address serialization so let's add the hook suffix for clarity

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 7]
```diff
@@ -107,6 +109,9 @@ export class HyperlaneIsmFactory extends HyperlaneApp<FactoryFactories> {
     } else if (config.type === ModuleType.AGGREGATION) {
       this.logger(`Deploying Aggregation ISM to ${chain}`);
       contract = await this.deployAggregationIsm(chain, config, origin);
+    } else if (config.type === ModuleType.OP_STACK) {
+      this.logger(`Deploying Aggregation ISM to ${chain}`);
```
Comment (yorhodes) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts:113]:
```suggestion
      this.logger(`Deploying Op Stack ISM to ${chain} for verifying ${origin}`);
```

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts (hunk 3 above)
Comment (yorhodes) [reply typescript/sdk/src/hook/HyperlaneHookDeployer.ts:158]: can you also pass the origin chain (iirc it's an optional arg)

File: typescript/sdk/src/hook/types.ts (hunk 6 above)
Comment (yorhodes) [inline typescript/sdk/src/hook/types.ts:13]:
```suggestion
  OP_STACK = 'opStackHook',
```

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 8]
```diff
@@ -214,6 +219,18 @@ export class HyperlaneIsmFactory extends HyperlaneApp<FactoryFactories> {
...
+  private async deployOpStackIsm(chain: ChainName, config: OpStackIsmConfig) {
+    const signer = this.multiProvider.getSigner(chain);
+    const overrides = this.multiProvider.getTransactionOverrides(chain);
+
+    const deployedIsm = await this.multiProvider.handleDeploy(
+      chain,
+      new OPStackIsm__factory(signer),
+      [config.nativeBridge, overrides],
+    );
+    return deployedIsm;
```
Comment (yorhodes) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts:231]: signer and overrides stuff is done in `handleDeploy`
```suggestion
    return this.multiProvider.handleDeploy(
      chain,
      new OPStackIsm__factory(),
      [config.nativeBridge],
    );
```

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 9]
```diff
@@ -355,6 +372,10 @@ export async function moduleCanCertainlyVerify(
         }
         return verified >= destModule.threshold;
       }
+      case ModuleType.OP_STACK:
+        return destModule.nativeBridge !== ethers.constants.AddressZero;
+      default:
+        throw new Error(`Unsupported module type: ${(destModule as any).type}`);
```
Comment (yorhodes) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts:378]: nice

File: typescript/sdk/src/ism/types.ts [hunk 10]
```diff
@@ -48,8 +51,14 @@ export type AggregationIsmConfig = {
   threshold: number;
 };
 
+export type OpStackIsmConfig = {
+  type: ModuleType.OP_STACK;
+  nativeBridge: Address;
```
Comment (yorhodes) [inline typescript/sdk/src/ism/types.ts:56]: I thought this authenticated the hook address

File: typescript/sdk/src/hook/HyperlaneHookDeployer.ts (hunk 4 above)
Comment (aroralanuk) [reply typescript/sdk/src/hook/HyperlaneHookDeployer.ts:154]: ismConfig doesn't need to have hook address bc setAuthorizedHook is after the hook deployment

File: typescript/sdk/src/hook/types.ts (hunk 6 above)
Comment (aroralanuk) [reply typescript/sdk/src/hook/types.ts:13]: done

File: typescript/sdk/src/ism/types.ts (hunk 10 above)
Comment (aroralanuk) [reply typescript/sdk/src/ism/types.ts:56]: added setAuthorizedHook to the end of deployOpStack(), the ismConfig doesn't need a hook address

**Pull request closed without merging**